### PR TITLE
feat(graph): Create new API endpoint metrics/download 

### PR DESCRIPTION
**Pull request by TamazC**
## Description

This PR adds new API endpoint metrics/download ins order to download graphs data as CSV file

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [x] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software

## Target serie

- [ ] 21.04.x
- [ ] 21.10.x
- [ ] 22.04.x
- [x] 22.10.x (master)

<h2> How this pull request can be tested </h2>

the API is available at [{$HOST}/centreon/api/latest/monitoring/hosts/{$HOST_ID}/services/{$SERVICE_ID}/metrics/download?start_date={date_in_ISO_8601_format}&end_date={date_in_ISO_8601_format}](url)

#### Community contributors & Centreon team

- [x] I have followed the **coding style guidelines** provided by Centreon
- [x] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [x] I have commented my code, especially **hard-to-understand areas** of the PR.
- [x] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

File: src/Core/Application/RealTime/Repository/ReadDataBinRepositoryInterface.php [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+use \DateTimeInterface ;
+
+interface ReadDataBinRepositoryInterface
+{
+    /**
+     * @param array<int, string> $metrics
+     * @param DateTimeInterface $startDate
+     * @param DateTimeInterface $endDate
+     * @return iterable<String[]>
```
Comment (TamazC) [inline src/Core/Application/RealTime/Repository/ReadDataBinRepositoryInterface.php]: iterable<array<string,string>>

File: src/Core/Application/RealTime/UseCase/FindPerformanceMetrics/FindPerformanceMetrics.php [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+ *
+ */
+
+declare(strict_types=1);
+
+namespace Core\Application\RealTime\UseCase\FindPerformanceMetrics;
+
+use \DateTimeInterface;
+use Centreon\Domain\Log\LoggerTrait;
+use Core\Application\RealTime\Repository\ReadDataBinRepositoryInterface;
```
Comment (TamazC) [inline src/Core/Application/RealTime/UseCase/FindPerformanceMetrics/FindPerformanceMetrics.php]: ReadPerformanceDataRepositoryInterface

File: src/Core/Infrastructure/RealTime/Api/DownloadPerformanceMetrics/DownloadPerformanceMetricsController.php [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+        $this->denyAccessUnlessGrantedForApiRealtime();
+
+        $this->request = $request;
+        $index = $indexDataRepository->findIndexByHostIdAndServiceId($hostId, $serviceId);
+        $metrics = $metricRepository->findMetricsByIndexId($index);
+        $this->findStartDate();
+        $this->findEndDate();
+        $fileName = $this->generateDownloadFileNameByIndex($index, $indexDataRepository);
+
+        return $useCase($metrics, $this->startDate, $this->endDate, $fileName);
```
Comment (TamazC) [inline src/Core/Infrastructure/RealTime/Api/DownloadPerformanceMetrics/DownloadPerformanceMetricsController.php]: use usecase instead

File: src/Core/Infrastructure/RealTime/Repository/FindIndex/DbReadIndexDataRepository.php [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+    public function findHostNameAndServiceDescriptionByIndex(int $index): array
+    {
+        $query = 'SELECT host_name, service_description FROM `:dbstg`.index_data WHERE id = :index';
+        $statement = $this->db->prepare($this->translateDbName($query));
+        $statement->bindValue(':index', $index, \PDO::PARAM_INT);
+
+        $statement->execute();
+
+        return $statement->fetch();
```
Comment (TamazC) [inline src/Core/Infrastructure/RealTime/Repository/FindIndex/DbReadIndexDataRepository.php]: Fetch against an object

Handle when false is returned

File: src/Core/Infrastructure/RealTime/Repository/FindIndex/DbReadIndexDataRepository.php [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+        $row = $statement->fetch();
+
+        if (!is_array($row) || !array_key_exists('id', $row)) {
+            throw new \InvalidArgumentException('Resource not found');
+        }
+
+        return (int) $row['id'];
+    }
+
+    public function findHostNameAndServiceDescriptionByIndex(int $index): array
```
Comment (TamazC) [inline src/Core/Infrastructure/RealTime/Repository/FindIndex/DbReadIndexDataRepository.php]: @inheritdoc

File: src/Core/Infrastructure/RealTime/Repository/FindIndex/DbReadIndexDataRepository.php [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+{
+    /**
+     * @param DatabaseConnection $db
+     */
+    public function __construct(DatabaseConnection $db)
+    {
+        $this->db = $db;
+    }
+
+    public function findIndexByHostIdAndServiceId(int $hostId, int $serviceId): int
```
Comment (TamazC) [inline src/Core/Infrastructure/RealTime/Repository/FindIndex/DbReadIndexDataRepository.php:45]: @inheritdoc

File: src/Core/Infrastructure/RealTime/Repository/FindMetric/DbReadMetricRepository.php [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+
+    public function findMetricsByIndexId(int $indexId): array
+    {
+        $query = 'SELECT DISTINCT metric_id, metric_name FROM `:dbstg`.metrics, `:dbstg`.index_data ';
+        $query .= ' WHERE metrics.index_id = index_data.id AND id = :index_id ORDER BY metric_id';
+        $statement = $this->db->prepare($this->translateDbName($query));
+        $statement->bindValue(':index_id', $indexId, \PDO::PARAM_INT);
+        $statement->execute();
+
+        return $statement->fetchAll(\PDO::FETCH_KEY_PAIR);
```
Comment (TamazC) [inline src/Core/Infrastructure/RealTime/Repository/FindMetric/DbReadMetricRepository.php]: fetch against an object

File: tests/php/Core/Application/RealTime/UseCase/FindPerformanceMetrics/FindPerformanceMetricResponseTest.php [hunk 8]
```diff
@@ -0,0 +1,118 @@
+<?php
```
Comment (TamazC) [inline tests/php/Core/Application/RealTime/UseCase/FindPerformanceMetrics/FindPerformanceMetricResponseTest.php:1]: migrate to pest

File: tests/php/Core/Application/RealTime/UseCase/FindPerformanceMetrics/FindPerformanceMetricsTest.php [hunk 9]
```diff
@@ -0,0 +1,218 @@
+<?php
```
Comment (TamazC) [inline tests/php/Core/Application/RealTime/UseCase/FindPerformanceMetrics/FindPerformanceMetricsTest.php:1]: migrate to pest

File: src/Core/Infrastructure/Common/Presenter/DownloadPresenter.php [hunk 10]
```diff
@@ -0,0 +1,33 @@
+<?php
```
Comment (TamazC) [inline src/Core/Infrastructure/Common/Presenter/DownloadPresenter.php:1]: generate file extension automaticly

File: src/Core/Infrastructure/Common/Presenter/DownloadPresenter.php [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+
```
Comment (TamazC) [inline src/Core/Infrastructure/Common/Presenter/DownloadPresenter.php:2]: Add licence

**Review by jeremyjaouen — DISMISSED** [review]

**Review by callapa — APPROVED** [review]

**Review by jeremyjaouen — APPROVED** [review]

**Pull request merged**